docs/advanced-features/react-18/streaming.md

```diff
@@ -0,0 +1,71 @@
+# Streaming SSR (Alpha)
+
+React 18 will include architectural improvements to React server-side rendering (SSR) performance. This means you can use `Suspense` in your React components in streaming SSR mode and React will render them on the server and send them through HTTP streams.
+It's worth noting that another experimental feature, React Server Components, is based on streaming. You can read more about server components related streaming APIs in [`next/streaming`](docs/api-reference/next/streaming.md). However, this guide focuses on basic React 18 streaming.
+
+## Enable Streaming SSR
+
+Enabling streaming SSR means React renders your components into streams and the client continues receiving updates from these streams even after the initial SSR response is sent. In other words, when any suspended components resolve down the line, they are rendered on the server and streamed to the client. With this strategy, the app can start emitting HTML even before all the data is ready, improving your app's loading performance. As an added bonus, in streaming SSR mode, the client will also use selective hydration strategy to prioritize component hydration which based on user interaction.
+
+To enable streaming SSR, set the experimental option `runtime` to either `'nodejs'` or `'edge'`:
+
+```jsx
+// next.config.js
+module.exports = {
+  experimental: {
+    runtime: 'nodejs',
+  },
+}
+```
+
+This option determines the environment in which streaming SSR will be happening. When setting to `'edge'`, the server will be running entirely in the [Edge Runtime](https://nextjs.org/docs/api-reference/edge-runtime).
+
+## Streaming Features
+
+### next/dynamic
+
+Dynamic imports through `React.lazy` have better support in React 18. Previously, Next.js supported dynamic imports internally without requiring `Suspense` or `React.lazy`. Now to embrace the official APIs on the React side, we provide you with `options.suspense` in `next/dynamic`.
+
+```jsx
+import dynamic from 'next/dynamic'
+import { lazy, Suspense } from 'react'
+
+import Content from '../components/content'
+
+// These two ways are identical:
```

Comment:
Not strictly identical, as `React.lazy` doesn't do any preloading. Might be worth suggesting that you continue using `next/dynamic` for now for performance?

Reply:
When we enable `{ suspense: true }` it's using `React.lazy` and doesn't do preloading I think?

Reply:
Oh, I see the issue -- we just haven't implemented the preloading yet, but we should. Otherwise there's no reason to have the `{ suspense: true }` option instead of just recommending that people use `React.lazy` instead.

Reply:
The need for `{suspense: true}` makes it feel second-class. Do we plan to flip the default at some point in the future?

Reply:
The issue is that currently that would be a breaking change without an incremental upgrade path: your application using `next/dynamic` would break when upgrading to React 18 unless you add suspense boundaries first, but you can't add suspense boundaries until you upgrade to React 18. It also changes the semantics a bit, since pre-Suspense, the loading state is associated with the `next/dynamic` component, rather than the parent.

Changing the default in the future is a possibility, but it's likely a major version away. Some shorter term options might be:
* If we could detect that we're rendering inside a `<Suspense>` boundary, we could assume it's intentional
* If we could somehow port the preloading behavior to `React.lazy`, we could push users toward that instead

Reply:
It occurred to me that with #23304 and #23267 this might not be a breaking change though, if the transition semantics apply to `root.render()` as well...

Reply:
I think what we could theoretically do is use suspense by default during SSR and initial hydration if you're using React>=18, since that behavior shouldn't change.

For client-side navigations that would still be at risk of suspending, we could render the legacy non-Suspense based fallback that you provide with `next/dynamic`, and then also warn that you should add a suspense boundary and move your fallback there instead.

Let me think about it some more.

Reply:
Just noticed that https://github.com/facebook/react/pull/23304 does in fact change the behavior for `root.render(...)` as long as it's not wrapped in `flushSync`. So yeah, it would probably make sense to make this change.

Thanks for flagging @gaearon